// @superfaceai/one-sdk/dist/internal/interpreter/sandbox/stdlib.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from) {
    for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
        to[j] = from[i];
    return to;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStdlib = void 0;
var debug_1 = __importDefault(require("debug"));
var object_1 = require("../../../lib/object");
var debugLog = debug_1.default('superface:debug-log');
var STDLIB_UNSTABLE = {
    time: {
        isoDateToUnixTimestamp: function (iso) {
            return new Date(iso).getTime();
        },
        unixTimestampToIsoDate: function (unix) {
            return new Date(unix).toISOString();
        },
    },
    debug: {
        log: function (formatter) {
            var args = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                args[_i - 1] = arguments[_i];
            }
            return debugLog.apply(void 0, __spreadArray([formatter], __read(args)));
        },
    },
};
var STDLIB = object_1.deepFreeze({
    unstable: STDLIB_UNSTABLE,
});
function getStdlib() {
    // TODO: This should later decide whether to return debug functions or just their stubs
    return STDLIB;
}
exports.getStdlib = getStdlib;
